File: app/src/main/java/com/example/floating/bakingapp/fragments/RecipeDetailsFragment.java
package com.example.floating.bakingapp.fragments;

import android.annotation.SuppressLint;
import android.content.res.Configuration;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.media.session.PlaybackStateCompat;
import android.support.v7.widget.CardView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.example.floating.bakingapp.R;
import com.example.floating.bakingapp.data.Steps;
import com.google.android.exoplayer2.DefaultLoadControl;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.extractor.DefaultExtractorsFactory;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.google.android.exoplayer2.ui.SimpleExoPlayerView;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.util.Util;

import java.util.ArrayList;

import static com.example.floating.bakingapp.ui.RecipeListActivity.steps;

/**
 * Created by Abdulkarim on 6/14/2017.
 */

public class RecipeDetailsFragment extends Fragment implements ExoPlayer.EventListener {

    /**
     * The fragment argument representing the item ID that this fragment
     * represents.
     */
    private static final String TAG = RecipeDetailsFragment.class.getSimpleName();
    public static final String ITEM_ID = "item_id";
    public static final String POSITION = "position";
    public static final String RECIPE_STEPS = "recipe_steps";
    public static final String PANES = "panes";
    public static final String STEPS = "steps";
    public static final String AUTOPLAY = "autoplay";
    public static final String CURRENT_WINDOW = "current_window";
    public static final String PLAYBACK_POSITION = "playback_position";


    public static ArrayList<Steps> steps_list;
    public static int mIndex = 0;
    private long mPosition;
    private boolean autoplay = false;
    public static boolean mTwoPane;
    View mStepDetails;
    int currVideoIndex;
    private SimpleExoPlayer mExoPlayer;
    private TextView mStepHeaderTextView, mStepDescriptionTextView;
    private SimpleExoPlayerView mPlayerView;
    private CardView details;


    /**
     * The dummy content this fragment is presenting.
     */
    //private DummyContent.DummyItem mItem;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */


    public RecipeDetailsFragment() {
    }



    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        if (savedInstanceState != null) {
            mIndex = savedInstanceState.getInt(ITEM_ID);
            mTwoPane = savedInstanceState.getBoolean(PANES);
            mPosition = savedInstanceState.getLong(POSITION,0);
            steps_list = savedInstanceState.getParcelableArrayList(STEPS);
            autoplay = savedInstanceState.getBoolean(AUTOPLAY, false);
            currVideoIndex = savedInstanceState.getInt(CURRENT_WINDOW, 0);
        }else {
            mIndex = getArguments().getInt(ITEM_ID);
            mTwoPane = getArguments().getBoolean(PANES);
            steps_list = getArguments().getParcelableArrayList(RECIPE_STEPS);
        }

        View rootView = inflater.inflate(R.layout.recipe_detail_fragment, container, false);

        mStepDetails = (View) rootView.findViewById(R.id.recipe_step_details_container);
        mPlayerView = (SimpleExoPlayerView) rootView.findViewById(R.id.video_player);
        mStepDescriptionTextView = (TextView) rootView.findViewById(R.id.recipe_step_description);
        mStepHeaderTextView = (TextView) rootView.findViewById(R.id.recipe_step_header);
        details = (CardView) rootView.findViewById(R.id.description);

        return rootView;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        releasePlayer();
    }

    @Override
    public void onPause() {
        super.onPause();
        releasePlayer();
    }

    @Override
    public void onResume() {
        super.onResume();
        showViewsPhone();
    }

    @Override
    public void onStop() {
        super.onStop();
        releasePlayer();
    }

    private void initializePlayer(Uri mediaUri) {
        mExoPlayer = null;
        mExoPlayer = ExoPlayerFactory.newSimpleInstance(
                getContext(),
                new DefaultTrackSelector(),
                new DefaultLoadControl());

        mExoPlayer.addListener(this);
        mPlayerView.setPlayer(mExoPlayer);
        mExoPlayer.setPlayWhenReady(true);

        mExoPlayer.seekTo(currVideoIndex, mPosition);

        String userAgent = Util.getUserAgent(getContext(), "The Baking App");

        MediaSource mediaSource = new ExtractorMediaSource(mediaUri,
                new DefaultDataSourceFactory(
                        getActivity(), userAgent),
                new DefaultExtractorsFactory(), null, null);

        mExoPlayer.prepare(mediaSource);
    }

    void showViewsPhone() {
        if (!mTwoPane && isLandscape()) {
            hideSystemUI();
            details.setVisibility(View.GONE);
            playInstructionVideo(mIndex);
            mStepHeaderTextView.setText(steps.get(mIndex).getShortDescription());
            mStepDescriptionTextView.setText(steps.get(mIndex).getDescription());
        } else {
            details.setVisibility(View.VISIBLE);
            mStepHeaderTextView.setText(steps.get(mIndex).getShortDescription());
            mStepDescriptionTextView.setText(steps.get(mIndex).getDescription());
            playInstructionVideo(mIndex);
        }
    }

    private void playInstructionVideo(int index) {
        mPlayerView.setVisibility(View.VISIBLE);
        mPlayerView.requestFocus();
        String thumbnailUrl = steps.get(index).getThumbnailURL();
        String videoUrl = steps.get(index).getVideoURL();

        if (!videoUrl.isEmpty()) {
            initializePlayer(Uri.parse(videoUrl));
        } else if (!thumbnailUrl.isEmpty()) {
            initializePlayer(Uri.parse(thumbnailUrl));
        }else {
            mPlayerView.setVisibility(View.GONE);
        }
    }

    private void releasePlayer() {
        if (mExoPlayer != null) {
            mPosition = mExoPlayer.getCurrentPosition();
            currVideoIndex = mExoPlayer.getCurrentWindowIndex();
            mExoPlayer.stop();
            mExoPlayer.release();
            mExoPlayer = null;
        }
    }


    boolean isLandscape() {
        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE)
            return true;
        return false;
    }

    @SuppressLint("InlineApi")
    private void hideSystemUI() {
        View view = getActivity().getWindow().getDecorView();

        view.setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_IMMERSIVE);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        if (mExoPlayer != null){
            outState.putLong(PLAYBACK_POSITION, mPosition);
            outState.putInt(CURRENT_WINDOW, currVideoIndex);
            outState.putBoolean(AUTOPLAY, autoplay);
        }
        outState.putBoolean(PANES, mTwoPane);
        outState.putInt(ITEM_ID, mIndex);
        outState.putParcelableArrayList(STEPS, steps_list);
    }

    @Override
    public void onTimelineChanged(Timeline timeline, Object manifest) {

    }

    @Override
    public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {

    }

    @Override
    public void onLoadingChanged(boolean isLoading) {

    }

    @Override
    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
        if (playbackState == ExoPlayer.STATE_READY && playWhenReady){
            Toast.makeText(getActivity(), "Playing", Toast.LENGTH_LONG).show();
        }
        if (playbackState == PlaybackStateCompat.STATE_PLAYING) {
            mPosition = mExoPlayer.getCurrentPosition();
        }
    }

    @Override
    public void onPlayerError(ExoPlaybackException error) {

    }

    @Override
    public void onPositionDiscontinuity() {

    }

    @Override
    public void onPlaybackParametersChanged(PlaybackParameters playbackParameters) {

    }
}

File: app/src/main/java/com/example/floating/bakingapp/details/fragment/RecipeDetailsFragment.kt
package com.example.floating.bakingapp.details.fragment

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.support.v4.app.Fragment
import android.support.v4.media.session.PlaybackStateCompat
import android.support.v7.widget.CardView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView

import com.example.floating.bakingapp.R
import com.example.floating.bakingapp.model.Steps
import com.example.floating.bakingapp.recipes.RecipeListActivity.Companion.steps
import com.google.android.exoplayer2.DefaultLoadControl
import com.google.android.exoplayer2.ExoPlaybackException
import com.google.android.exoplayer2.ExoPlayer
import com.google.android.exoplayer2.ExoPlayerFactory
import com.google.android.exoplayer2.PlaybackParameters
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.Timeline
import com.google.android.exoplayer2.extractor.DefaultExtractorsFactory
import com.google.android.exoplayer2.source.ExtractorMediaSource
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.google.android.exoplayer2.ui.SimpleExoPlayerView
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.util.Util

import org.parceler.Parcels

import java.util.ArrayList

/**
 * Created by Abdulkarim on 6/14/2017.
 */

/**
 * The dummy content this fragment is presenting.
 */
//private DummyContent.DummyItem mItem;

/**
 * Mandatory empty constructor for the fragment manager to instantiate the
 * fragment (e.g. upon screen orientation changes).
 */


class RecipeDetailsFragment : Fragment(), ExoPlayer.EventListener {
    private var mPosition: Long = 0
    private var autoplay = false
    internal lateinit var mStepDetails: View
    internal var currVideoIndex: Int = 0
    private var mExoPlayer: SimpleExoPlayer? = null
    private var mStepHeaderTextView: TextView? = null
    private var mStepDescriptionTextView: TextView? = null
    private var progressBar: ProgressBar? = null
    private var mPlayerView: SimpleExoPlayerView? = null
    private var details: CardView? = null


    internal val isLandscape: Boolean
        get() = resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        if (savedInstanceState != null) {
            mIndex = savedInstanceState.getInt(ITEM_ID)
            mTwoPane = savedInstanceState.getBoolean(PANES)
            mPosition = savedInstanceState.getLong(POSITION, 0)
            steps_list = savedInstanceState.getParcelableArrayList(STEPS)
            autoplay = savedInstanceState.getBoolean(AUTOPLAY, false)
            currVideoIndex = savedInstanceState.getInt(CURRENT_WINDOW, 0)
        } else {
            mIndex = arguments!!.getInt(ITEM_ID)
            mTwoPane = arguments!!.getBoolean(PANES)
            steps_list = arguments!!.getParcelableArrayList(RECIPE_STEPS)
        }

        val rootView = inflater.inflate(R.layout.fragment_recipe_details, container, false)

        mStepDetails = rootView.findViewById(R.id.recipe_step_details_container)
        mPlayerView = rootView.findViewById(R.id.video_player)
        progressBar = rootView.findViewById(R.id.progress_bar)
        mStepDescriptionTextView = rootView.findViewById(R.id.recipe_step_description)
        mStepHeaderTextView = rootView.findViewById(R.id.recipe_step_header)
        details = rootView.findViewById(R.id.description)

        return rootView
    }

    override fun onDestroy() {
        super.onDestroy()
        releasePlayer()
    }

    override fun onPause() {
        super.onPause()
        releasePlayer()
    }

    override fun onResume() {
        super.onResume()
        showViewsPhone()
    }

    override fun onStop() {
        super.onStop()
        releasePlayer()
    }

    private fun initializePlayer(mediaUri: Uri) {
        mExoPlayer = null
        mExoPlayer = ExoPlayerFactory.newSimpleInstance(
                context,
                DefaultTrackSelector(),
                DefaultLoadControl())

        mExoPlayer!!.addListener(this)
        mPlayerView!!.player = mExoPlayer
        mExoPlayer!!.playWhenReady = true

        mExoPlayer!!.seekTo(currVideoIndex, mPosition)

        val userAgent = Util.getUserAgent(context, "The Baking App")

        val mediaSource = ExtractorMediaSource(mediaUri,
                DefaultDataSourceFactory(
                        activity!!, userAgent),
                DefaultExtractorsFactory(), null, null)

        mExoPlayer!!.prepare(mediaSource)
    }

    internal fun showViewsPhone() {
        if (!mTwoPane && isLandscape) {
            hideSystemUI()
            details!!.visibility = View.GONE
            playInstructionVideo(mIndex)
            mStepHeaderTextView!!.text = steps[mIndex].shortDescription
            mStepDescriptionTextView!!.text = steps[mIndex].description
        } else {
            details!!.visibility = View.VISIBLE
            mStepHeaderTextView!!.text = steps[mIndex].shortDescription
            mStepDescriptionTextView!!.text = steps[mIndex].description
            playInstructionVideo(mIndex)
        }
    }

    private fun playInstructionVideo(index: Int) {
        mPlayerView!!.visibility = View.VISIBLE
        mPlayerView!!.requestFocus()
        val thumbnailUrl = steps[index].thumbnailURL
        val videoUrl = steps[index].videoURL

        if (!videoUrl!!.isEmpty()) {
            initializePlayer(Uri.parse(videoUrl))
        } else if (!thumbnailUrl!!.isEmpty()) {
            initializePlayer(Uri.parse(thumbnailUrl))
        } else {
            mPlayerView!!.visibility = View.GONE
        }
    }

    private fun releasePlayer() {
        if (mExoPlayer != null) {
            mPosition = mExoPlayer!!.currentPosition
            currVideoIndex = mExoPlayer!!.currentWindowIndex
            mExoPlayer!!.stop()
            mExoPlayer!!.release()
            mExoPlayer = null
        }
    }

    @SuppressLint("InlineApi")
    private fun hideSystemUI() {
        val view = activity!!.window.decorView

        view.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN
                or View.SYSTEM_UI_FLAG_IMMERSIVE)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        if (mExoPlayer != null) {
            outState.putLong(PLAYBACK_POSITION, mPosition)
            outState.putInt(CURRENT_WINDOW, currVideoIndex)
            outState.putBoolean(AUTOPLAY, autoplay)
        }
        outState.putBoolean(PANES, mTwoPane)
        outState.putInt(ITEM_ID, mIndex)
        outState.putParcelable(STEPS, Parcels.wrap(steps_list))
    }

    override fun onTimelineChanged(timeline: Timeline, manifest: Any) {

    }

    override fun onTracksChanged(trackGroups: TrackGroupArray, trackSelections: TrackSelectionArray) {

    }

    override fun onLoadingChanged(isLoading: Boolean) {

    }

    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
        if (playbackState == Player.STATE_READY && playWhenReady) {
            progressBar!!.visibility = View.GONE
        } else if (playbackState == Player.STATE_BUFFERING && playWhenReady) {
            progressBar!!.visibility = View.VISIBLE
        } else if (playbackState == PlaybackStateCompat.STATE_PLAYING) {
            mPosition = mExoPlayer!!.currentPosition
        }
    }

    override fun onRepeatModeChanged(repeatMode: Int) {

    }

    override fun onShuffleModeEnabledChanged(shuffleModeEnabled: Boolean) {

    }

    override fun onPlayerError(error: ExoPlaybackException) {

    }

    override fun onPositionDiscontinuity(reason: Int) {

    }

    override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {

    }

    override fun onSeekProcessed() {

    }

    companion object {

        /**
         * The fragment argument representing the item ID that this fragment
         * represents.
         */
        private val TAG = RecipeDetailsFragment::class.java.simpleName
        val ITEM_ID = "item_id"
        val POSITION = "position"
        val RECIPE_STEPS = "recipe_steps"
        val PANES = "panes"
        val STEPS = "steps"
        val AUTOPLAY = "autoplay"
        val CURRENT_WINDOW = "current_window"
        val PLAYBACK_POSITION = "playback_position"


        lateinit var steps_list: ArrayList<Steps>
        var mIndex = 0
        var mTwoPane: Boolean = false
    }
}
